aip/general/0128.md

```diff
@@ -0,0 +1,106 @@
+---
+id: 128
+state: reviewing
+created: 2020-09-14
+placement:
+  category: resource-design
+  order: 65
+---
+
+# Declarative-friendly interfaces
+
+Many services need to interact with common DevOps tools, particularly those
+that create and manage network-addressible resources (such as virtual machines,
+load balancers, database instances, and so on). These tools revolve around the
+principle of "configuration as code": the user specifies the complete intended
+landscape, and tooling is responsible for making whatever changes are necessary
+to achieve the user's specification.
+
+These tools are **declarative**: rather than specifying specific _actions_ to
+take, they specify the desired _outcome_, with the actions being derived based
+on the differences between the current landscape and the intended one.
+
+Furthermore, there are numerous popular DevOps tools, with more being
+introduced each year. Integrating hundreds of resource types with multiple
+tools requires strong consistency, so that integration can be automated.
+
+## Guidance
+
+Services **should** clearly delineate between "control plane" operations and
+"data plane" operations, ideally through the use of distinct services with
+their own interface definition documents.
+
+- Control plane operations are responsible for managing the _lifecycle_ of
+  resources.
+- Data plane operations are responsible for managing the _content_ of
+  resources.
+
+The same resource **may** have both control plane operations and data plane
+operations associated with it. For example, a database API would have
+operations to create or delete database tables (control plane) as well as
+operations to write and read rows to that table (data plane).
+
+### Resources
+
+Resources that are declarative-friendly **must** use only strongly-consistent
+standard methods for managing resource lifecycle, which allows tools to support
+these resources generically, as well as conforming to other
+declarative-friendly guidance (see [further reading](#further-reading)).
+
+Declarative-friendly resources **should** designate that they follow the
+declarative-friendly style:
+
+```proto
+message Book {
+  option (google.api.resource) = {
+    type: "library.googleapis.com/Book"
+    pattern: "publishers/{publisher}/books/{book}"
+    style: DECLARATIVE_FRIENDLY
+  };
+
+  string name = 1;
+  // Other fields...
+}
+```
+
+### Annotations
```

Comment:
We should delay merging this until the discussion on the internal doc concludes.